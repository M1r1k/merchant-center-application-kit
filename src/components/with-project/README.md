# `WithProject`

This is a data fetching component to fetch a project.
You can use this component (or the complementary HoC) to retrieve project data and use it when needed.

The data is cached within Apollo, so rendering the component multiple times won't trigger multiple requests but will load the data from the cache instead.

## `<WithProject>`

A React component that will pass project data to a `render` function.

> This is the React version of the `withProject` HoC.

### Usage

```js
import { WithProject } from '@commercetools-local/application-shell'

<WithProject
  projectKey="my-project"
  mapDataToProps={userData => userData}
  render={({ projectData: { loading, project }, ...rest }) =>
    loading
      ? <Loading />
      : (
        <div>
          <h1>{project.name}</h1>
          <Main {...rest} />
        </div>
      )
  }
/>
```

### Properties

| Props | Type | Required | Values | Default  | Description |
| --- | --- | :---: | --- | --- | --- |
| `projectKey` | `string` | ✅ | - | - | The `key` of the project that you want to fetch |
| `mapDataToProps` | `func` | - | - | - | Map the props that will be passed to the `render` function. Use this as a chance to compute some values within the data and pass it s a specific prop. |
| `render` | `func` | ✅ | - | - | Render your children elements within this function. The argument is an object containing the mapped data from `mapDataToProps` or the default bject injected by `graphql` HoC |

## `withProject(getProjectKey, mapDataToProps)`

A HoC component that will inject project data as prop.

> This is the HoC version of the `WithProject` component.

### Usage

```js
import { withProject } from '@commercetools-local/application-shell'

const ProjectTitle = props => (
  <div>{props.projectName}</div>
)
withProject(
  ownProps => ownProps.match.params.projectKey,
  projectData => ({ projectName: projectData.project && projectData.project.name })
)(ProjectTitle)
```

### Arguments

* `getProjectKey(ownProps): String`: a function that receives props coming from the parent component and returns the project key
* `mapDataToProps(projectData): Object`: map the props that will be injected to the composed component. Use this as a chance to compute some values within the data and pass it as a specific prop.
