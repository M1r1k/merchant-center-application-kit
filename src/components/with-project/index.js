export { default, withProject } from './with-project';
