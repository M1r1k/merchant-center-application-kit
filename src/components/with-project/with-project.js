import React from 'react';
import PropTypes from 'prop-types';
import { compose, setDisplayName } from 'recompose';
import gql from 'graphql-tag';
import { graphql } from 'react-apollo';
import omit from 'lodash.omit';

const ProjectQuery = gql`
  query Project($projectKey: String!) {
    project(key: $projectKey) {
      id
      key
      name
      countries
      currencies
      languages
      expired
      suspended
      permissions {
        canManageOrganization
        canManageProject
        canViewProjectSettings
        canViewProducts
        canManageProducts
        canViewOrders
        canManageOrders
        canViewCustomers
        canManageCustomers
        canViewTypes
        canManageTypes
        canViewShippingLists
        canManageShippingLists
        canViewPayments
        canManagePayments
      }
      owner {
        id
        name
        createdAt
        teamsCount
      }
      settings {
        id
        baseSettings {
          id
          userProjectSettings
          pluginPreset
          imageRegex {
            thumb {
              ...ImageRegex
            }
            small {
              ...ImageRegex
            }
          }
          orderStatesVisibility {
            isPaymentStateHidden
            isShipmentStateHidden
            isOrderStateHidden
          }
          categoryRecommendationSettings {
            searchProperty
            options {
              attributeName
            }
          }
        }
        productSettings
        currentProductSettings
      }
    }
  }
`;

const graphqlOptions = {
  alias: 'withProject',
  name: 'projectData',
};

const WithProject = props => {
  const mappedProps = props.mapDataToProps
    ? props.mapDataToProps(props.projectData)
    : { projectData: props.projectData };
  // Extract props coming from parents and omit props specific to this
  // component, then pass them down to the children.
  const parentProps = omit(props, [
    'projectKey',
    'mapDataToProps',
    'render',
    'projectData',
  ]);
  return props.render({ ...parentProps, ...mappedProps });
};
WithProject.displayName = 'WithProject';
WithProject.propTypes = {
  projectKey: PropTypes.string.isRequired,
  mapDataToProps: PropTypes.func,
  render: PropTypes.func.isRequired,
  // Injected
  projectData: PropTypes.shape({
    loading: PropTypes.bool.isRequired,
    error: PropTypes.object,
    project: PropTypes.object, // see graphql query shape
  }).isRequired,
};

// React component
const WithUserProject = compose(
  setDisplayName('WithProject'),
  graphql(ProjectQuery, graphqlOptions)
)(WithProject);

// HoC
const withProject = (getProjectKey, mapDataToProps) => Component => {
  const WrappedWithProject = props => (
    <WithUserProject
      projectKey={getProjectKey(props)}
      mapDataToProps={mapDataToProps}
      render={mappedProps => <Component {...props} {...mappedProps} />}
    />
  );
  WrappedWithProject.displayName = 'WithProject';
  return WrappedWithProject;
};

// Public exports
export default WithUserProject;
export { withProject };

// For testing
export { ProjectQuery, WithProject };
