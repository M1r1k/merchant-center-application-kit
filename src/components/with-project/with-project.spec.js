import React from 'react';
import { shallow } from 'enzyme';
import { WithProject, withProject, ProjectQuery } from './with-project';

describe('rendering', () => {
  let props;
  let wrapper;

  describe('<WithProject>', () => {
    describe('when mapDataToProps is not defined', () => {
      beforeEach(() => {
        props = {
          projectKey: 'my-project',
          render: jest.fn(),
          // this is usually injected by graphql
          projectData: {
            loading: false,
            project: {
              name: 'My Project',
            },
          },
        };
        wrapper = shallow(<WithProject foo="bar" {...props} />);
      });
      it('should call render with projectData object', () => {
        expect(props.render).toHaveBeenCalledWith(
          expect.objectContaining({ projectData: props.projectData })
        );
      });
      it('should proxy parent props to render', () => {
        expect(props.render).toHaveBeenCalledWith(
          expect.objectContaining({ foo: 'bar' })
        );
      });
    });
    describe('when mapDataToProps is defined', () => {
      beforeEach(() => {
        props = {
          projectKey: 'my-project',
          mapDataToProps: projectData => ({
            name: projectData.project && projectData.project.name,
          }),
          render: jest.fn(),
          // this is usually injected by graphql
          projectData: {
            loading: false,
            project: {
              name: 'My Project',
            },
          },
        };
        wrapper = shallow(<WithProject foo="bar" {...props} />);
      });
      it('should call render with mapped prop name', () => {
        expect(props.render).toHaveBeenCalledWith(
          expect.objectContaining({ name: 'My Project' })
        );
      });
      it('should call render without projectData object', () => {
        expect(props.render).not.toHaveBeenCalledWith(
          expect.objectContaining({ projectData: props.projectData })
        );
      });
      it('should proxy parent props to render', () => {
        expect(props.render).toHaveBeenCalledWith(
          expect.objectContaining({ foo: 'bar' })
        );
      });
    });
  });

  describe('withProject()', () => {
    let wrapperRender;
    // eslint-disable-next-line react/prop-types
    const ProjectTitle = ({ name }) => <div>{name}</div>;

    beforeEach(() => {
      const ProfileTitleWithProject = withProject(
        ownProps => ownProps.projectKey
      )(ProjectTitle);
      wrapper = shallow(
        <ProfileTitleWithProject foo="bar" projectKey="my-project" />
      );
      wrapperRender = shallow(wrapper.prop('render')({ name: 'My Project' }));
    });
    it('should render <ProjectTitle> with name', () => {
      expect(wrapperRender).toMatchElement(<div>{'My Project'}</div>);
    });
    it('should render WithProject internally', () => {
      expect(wrapper).toRender('WithProject');
    });
  });

  describe('graphql query', () => {
    it('should match snapshot', () => {
      expect(ProjectQuery).toMatchSnapshot();
    });
  });
});
