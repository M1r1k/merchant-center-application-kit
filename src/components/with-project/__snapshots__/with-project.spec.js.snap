// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering graphql query should match snapshot 1`] = `
Object {
  "definitions": Array [
    Object {
      "directives": Array [],
      "kind": "OperationDefinition",
      "name": Object {
        "kind": "Name",
        "value": "Project",
      },
      "operation": "query",
      "selectionSet": Object {
        "kind": "SelectionSet",
        "selections": Array [
          Object {
            "alias": null,
            "arguments": Array [
              Object {
                "kind": "Argument",
                "name": Object {
                  "kind": "Name",
                  "value": "key",
                },
                "value": Object {
                  "kind": "Variable",
                  "name": Object {
                    "kind": "Name",
                    "value": "projectKey",
                  },
                },
              },
            ],
            "directives": Array [],
            "kind": "Field",
            "name": Object {
              "kind": "Name",
              "value": "project",
            },
            "selectionSet": Object {
              "kind": "SelectionSet",
              "selections": Array [
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "id",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "key",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "name",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "countries",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "currencies",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "languages",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "expired",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "suspended",
                  },
                  "selectionSet": null,
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "permissions",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canManageOrganization",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canManageProject",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canViewProjectSettings",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canViewProducts",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canManageProducts",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canViewOrders",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canManageOrders",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canViewCustomers",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canManageCustomers",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canViewTypes",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canManageTypes",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canViewShippingLists",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canManageShippingLists",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canViewPayments",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "canManagePayments",
                        },
                        "selectionSet": null,
                      },
                    ],
                  },
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "owner",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "name",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "createdAt",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "teamsCount",
                        },
                        "selectionSet": null,
                      },
                    ],
                  },
                },
                Object {
                  "alias": null,
                  "arguments": Array [],
                  "directives": Array [],
                  "kind": "Field",
                  "name": Object {
                    "kind": "Name",
                    "value": "settings",
                  },
                  "selectionSet": Object {
                    "kind": "SelectionSet",
                    "selections": Array [
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "id",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "baseSettings",
                        },
                        "selectionSet": Object {
                          "kind": "SelectionSet",
                          "selections": Array [
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "id",
                              },
                              "selectionSet": null,
                            },
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "userProjectSettings",
                              },
                              "selectionSet": null,
                            },
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "pluginPreset",
                              },
                              "selectionSet": null,
                            },
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "imageRegex",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "thumb",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "selections": Array [
                                        Object {
                                          "directives": Array [],
                                          "kind": "FragmentSpread",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "ImageRegex",
                                          },
                                        },
                                      ],
                                    },
                                  },
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "small",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "selections": Array [
                                        Object {
                                          "directives": Array [],
                                          "kind": "FragmentSpread",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "ImageRegex",
                                          },
                                        },
                                      ],
                                    },
                                  },
                                ],
                              },
                            },
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "orderStatesVisibility",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "isPaymentStateHidden",
                                    },
                                    "selectionSet": null,
                                  },
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "isShipmentStateHidden",
                                    },
                                    "selectionSet": null,
                                  },
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "isOrderStateHidden",
                                    },
                                    "selectionSet": null,
                                  },
                                ],
                              },
                            },
                            Object {
                              "alias": null,
                              "arguments": Array [],
                              "directives": Array [],
                              "kind": "Field",
                              "name": Object {
                                "kind": "Name",
                                "value": "categoryRecommendationSettings",
                              },
                              "selectionSet": Object {
                                "kind": "SelectionSet",
                                "selections": Array [
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "searchProperty",
                                    },
                                    "selectionSet": null,
                                  },
                                  Object {
                                    "alias": null,
                                    "arguments": Array [],
                                    "directives": Array [],
                                    "kind": "Field",
                                    "name": Object {
                                      "kind": "Name",
                                      "value": "options",
                                    },
                                    "selectionSet": Object {
                                      "kind": "SelectionSet",
                                      "selections": Array [
                                        Object {
                                          "alias": null,
                                          "arguments": Array [],
                                          "directives": Array [],
                                          "kind": "Field",
                                          "name": Object {
                                            "kind": "Name",
                                            "value": "attributeName",
                                          },
                                          "selectionSet": null,
                                        },
                                      ],
                                    },
                                  },
                                ],
                              },
                            },
                          ],
                        },
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "productSettings",
                        },
                        "selectionSet": null,
                      },
                      Object {
                        "alias": null,
                        "arguments": Array [],
                        "directives": Array [],
                        "kind": "Field",
                        "name": Object {
                          "kind": "Name",
                          "value": "currentProductSettings",
                        },
                        "selectionSet": null,
                      },
                    ],
                  },
                },
              ],
            },
          },
        ],
      },
      "variableDefinitions": Array [
        Object {
          "defaultValue": null,
          "kind": "VariableDefinition",
          "type": Object {
            "kind": "NonNullType",
            "type": Object {
              "kind": "NamedType",
              "name": Object {
                "kind": "Name",
                "value": "String",
              },
            },
          },
          "variable": Object {
            "kind": "Variable",
            "name": Object {
              "kind": "Name",
              "value": "projectKey",
            },
          },
        },
      ],
    },
  ],
  "kind": "Document",
  "loc": Object {
    "end": 1330,
    "start": 0,
  },
}
`;
